### PR TITLE
The autoscaler does not scale node groups on non-amd64 clusters when pods explicitly require non-amd64 nodes in node affinity

### LINKED ISSUE #6064: The autoscaler does not scale node groups on non-amd64 clusters when pods explicitly require non-amd64 nodes in node affinity
The autoscaler does not scale node groups on non-amd64 clusters when pods explicitly require non-amd64 nodes in node affinity

**Which component are you using?**: cluster-autoscaler

**What version of the component are you using?**: master branch

Component version: any

**What k8s version are you using (`kubectl version`)?**: any

**What environment is this in?**: cluster API and most cloud providers

**What did you expect to happen?**: when the autoscaler provider build the generic labels to perform autoscale from zero decisions, most cloud providers consider `amd64` as the default architecture to use, leading a pod reporting a node selector term to be scheduled on non-amd64 nodes, unable to trigger the autoscaler although non-amd64 nodes are available in the cluster.

What behaviour did you expect to see?

Users should be allowed to override the default architecture in the buildGenericLabels method. This will not take into account cases like clusters heterogeneous in architecture (CP on arm64, DP on amd64, or CP on amd64 and DP arm64 + amd64 nodes). However, this case should be considered user-specific or cloud-provider-specific.

**What happened instead?**:

When a pod reports a node selector term like `kubernetes.io/arch in {arm64}` in an arm64 cluster, the pod does not trigger the scale out of node groups.

**How to reproduce it (as minimally and precisely as possible)**:

In any kubernetes arm64 deployment with the autoscaler enabled, try to define a deployment with node selector term `kubernetes.io/arch in {arm64}` , resource requests, and set the replicas such that a pod keeps pending scheduling due to missing resources.

Example:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: 'my-deployment'
spec:
  selector:
    matchLabels:
      app: name
  replicas: 3
  template:
    metadata:
      labels:
        app: name
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                - key: kubernetes.io/arch
                  operator: In
                  values:
                    - "arm64"
      containers:
        - name: container
          image: >-
            quay.io/openshifttest/hello-openshift:1.2.0
          resources:
              requests:
                cpu: "2"
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
  paused: false
```

### PR DESCRIPTION
**Pull request by aleskandro**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

The architecture label in the build generic labels method of the Cluster API provider is now populated using the `cpu.GetDefaultScaleFromZeroArchitecture().Name()` method.

The method allows users deploying the cluster-autoscaler to define the default architecture to be used by the cluster-autoscaler for scale up from zero via the `--scale-up-from-zero-default-arch` flag. `Amd64` is kept as a fallback for historical reasons and to maintain compatibility with current providers.

The `SystemArchitecture` type and the functions for parsing architecture names are copied over from the GCE implementation, to be generic across all the defined cloud providers.

However, the GCE implementation and the other cloud providers are currently left untouched.

The introduced changes will not take into account the case of nodes heterogeneous in architecture. The labels generation to infer properties like the cpu architecture from the node groups' features should be considered cloud-provider/user specific.

Cloud providers willing to opt into this implementation are expected to change their manager code by replacing
the `cloudprovider.DefaultArch` constant with the `GetDefaultScaleFromZeroArchitecture` function exposed by this package.
Usually, this should be done in the function responsible for generating the generic labels of the given
cloud provider's manager code (for example, `buildGenericLabels(.)`) as follows:

```go
func (...) buildGenericLabels(...) (map[string]string) {
	result := make(map[string]string)
	result[apiv1.LabelArchStable] = cpu.GetDefaultScaleFromZeroArchitecture().Name()
	// ...
}
```

#### Which issue(s) this PR fixes:

Fixes #6064 

#### Does this PR introduce a user-facing change?

```release-note
Fixes #6064: Users of Cluster API can override the default architecture to consider in the templates for autoscaling from zero so that pods requesting non-amd64 nodes in their node selector terms can trigger the scale-up in non-amd64 single-arch clusters. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

cc @elmiko @jeffdyoung @Prashanth684

### CODE REVIEW

**Review by elmiko — COMMENTED** [review]
thanks for this update @aleskandro , in general this looks good to me and i'm happy to approve the cluster-api changes.

i am concerned about getting consensus on this pr since it touches so many other providers.

/lgtm

**Comment by elmiko** [comment]
we talked about this at the sig meeting today, the general consensus is that this should be limited to just cluster-api for now. we can add the flag to help determine the default, and provide some code comments for how other providers can use it. but to start, only cluster-api will respect the flag.

/lgtm cancel

**Comment by aleskandro** [comment]
Hi @elmiko , thanks for the review and thanks all for the discussion during the sig meeting today.

I made the requested changes:
- No other cloud providers than Cluster API are now affected by changes in this PR
- The SystemArchietcture type and its methods are now copied over from the GCE implementation (it is left untouched, too)
- Updated usage description and godoc comments with information about how to allow other cloud providers to opt in the facility given by the new command line argument
- Updated the CAPI readme to inform users about Arch-awareness in the scale from zero for single-arch clusters.

File: cluster-autoscaler/cloudprovider/clusterapi/README.md [hunk 1]
```diff
@@ -275,6 +276,16 @@ metadata:
     capacity.cluster-autoscaler.kubernetes.io/taints: "key1=value1:NoSchedule,key2=value2:NoExecute"
 ```
 
+#### CPU Architecture awareness for single-arch clusters 
+
+Users of non-amd64, single-arch clusters willing to enable scale from zero
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/README.md:281]: i think i would reword this as:

```suggestion
Users of single-arch non-amd64 clusters who are using scale from zero 
```

File: cluster-autoscaler/cloudprovider/clusterapi/README.md [hunk 2]
```diff
@@ -275,6 +276,16 @@ metadata:
     capacity.cluster-autoscaler.kubernetes.io/taints: "key1=value1:NoSchedule,key2=value2:NoExecute"
 ```
 
+#### CPU Architecture awareness for single-arch clusters 
+
+Users of non-amd64, single-arch clusters willing to enable scale from zero 
+support should also set the `--scale-up-from-zero-default-arch` flag to the 
+architecture of the nodes they want to default the node group templates to.
+The autoscaler will default to amd64 if the flag is not set, and the node
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/README.md:284]: i would add some code blocks here for style

```suggestion
The autoscaler will default to `amd64` if the flag is not set, and the node 
```

File: cluster-autoscaler/cloudprovider/clusterapi/README.md [hunk 3]
```diff
@@ -275,6 +276,16 @@ metadata:
     capacity.cluster-autoscaler.kubernetes.io/taints: "key1=value1:NoSchedule,key2=value2:NoExecute"
 ```
 
+#### CPU Architecture awareness for single-arch clusters 
+
+Users of non-amd64, single-arch clusters willing to enable scale from zero 
+support should also set the `--scale-up-from-zero-default-arch` flag to the 
+architecture of the nodes they want to default the node group templates to.
+The autoscaler will default to amd64 if the flag is not set, and the node 
+group templates will not match the nodes' architecture, particularly when
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/README.md:285]: minor change

```suggestion
group templates will not match the nodes' architecture, specifically when 
```

**Review by elmiko — COMMENTED** [review]
i think this is looking really nice @aleskandro , much easier to review. i also like that you updated the PR description, hopefully we could get more providers to adopt this logic.

i just have some suggestions for the readme, but otherwise i'm happy to approve this. we will also need to get approval for the flag changes and cpu module addition.

File: cluster-autoscaler/utils/cpu/architecture.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+// Cloud providers willing to opt into this implementation are expected to change their manager code by replacing
+// the cloudprovider.DefaultArch constant with the GetDefaultScaleFromZeroArchitecture function exposed by this package.
+// Usually, this should be done in the function responsible for generating the generic labels of the given
+// cloud provider's manager code (for example, buildGenericLabels(.)) as follows:
+//
+//	func (...) buildGenericLabels(...) (map[string]string) {
+//		result := make(map[string]string)
+//		result[apiv1.LabelArchStable] = cpu.GetDefaultScaleFromZeroArchitecture().Name()
+//		// ...
+//	}
```
Comment (elmiko) [inline cluster-autoscaler/utils/cpu/architecture.go:64]: :heart:

**Comment by aleskandro** [comment]
> i think this is looking really nice @aleskandro , much easier to review. i also like that you updated the PR description, hopefully we could get more providers to adopt this logic.
> 
> i just have some suggestions for the readme, but otherwise i'm happy to approve this. we will also need to get approval for the flag changes and cpu module addition.

Hi @elmiko , thanks again. PTAL when you get the bandwidth for it.

**Review by elmiko — COMMENTED** [review]
/lgtm

i like the comments in the new cpu util package, makes it clear how others can participate. if the sig is ok with adding the new flag i think this is great.

**Review by elmiko — COMMENTED** [review]
thanks @aleskandro 

/lgtm

**Review by elmiko — COMMENTED** [review]
oops, forgot to add approval from my side

/approve

we will need a core reviewer to approve the util stuff.

**Comment by BigDarkClown** [comment]
Really my only comment is that this feature seems *very* cloud provider specific. I'm not sure how useful it will be to other cloud providers (different architectures etc.), so I'm a bit iffy on placing this architecture defaulting logic in the shared space.

Additionally, the CA is following the practise of specifying all the component-wide flags in the `main.go`. I'm very far from arguing that this is in any way superior to the method you used (I actually think it's worse), but for the sake of consistency it would be nice to add the flag in already established space, just so it does not get lost in the future.

**Comment by elmiko** [comment]
> Really my only comment is that this feature seems very cloud provider specific. I'm not sure how useful it will be to other cloud providers (different architectures etc.), so I'm a bit iffy on placing this architecture defaulting logic in the shared space.

i think that's fair, we had initially thought that other providers might want to use the same defaulting, but if this should just be a clusterapi change then we could certainly use an environment variable (as we have done with other clusterapi specific config) instead of the flag. that would allow us to move all the code into the clusterapi directory.

good point about the flag.

**Comment by aleskandro** [comment]
> Additionally, the CA is following the practise of specifying all the component-wide flags in the main.go. I'm very far from arguing that this is in any way superior to the method you used (I actually think it's worse), but for the sake of consistency it would be nice to add the flag in already established space, just so it does not get lost in the future.

Yeah, I imagined this but we wanted to add it in the review discussion as adding the flag to the main and exposing the variable as public to all the packages is far from the ideal too IMHO.

> > Really my only comment is that this feature seems very cloud provider specific. I'm not sure how useful it will be to other cloud providers (different architectures etc.), so I'm a bit iffy on placing this architecture defaulting logic in the shared space.
> 
> i think that's fair, we had initially thought that other providers might want to use the same defaulting, but if this should just be a clusterapi change then we could certainly use an environment variable (as we have done with other clusterapi specific config) instead of the flag. that would allow us to move all the code into the clusterapi directory.
> 
> good point about the flag.

This issue should affect all the cloud providers but GCE currently. However yes, other providers could decide for a different way of fixing it. I hoped to provide a more general solution.

Changed the implementation to be cluster api specific and using an env var.

**Comment by elmiko** [comment]
thanks @aleskandro , that will make it easier to merge.

i'm fine to merge this PR as only cluster-api, but if you wanted to try and build a more general solution perhaps it would be appropriate to make a new [proposal](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/proposals) ?

then we could debate the implementation on the proposal, and we would have a central document about how to implement the change (even though i loved your code comments).

**Comment by aleskandro** [comment]
Ok @elmiko, would we merge this first and run the proposal as a follow up, keeping the issue open?

**Comment by elmiko** [comment]
good question @aleskandro , i think we should definitely merge this and then create the proposal. given that the issue is wider in scope than just clusterapi, i'd say yes we probably want to keep it open.

File: cluster-autoscaler/cloudprovider/clusterapi/README.md [hunk 5]
```diff
@@ -276,6 +277,16 @@ metadata:
     capacity.cluster-autoscaler.kubernetes.io/taints: "key1=value1:NoSchedule,key2=value2:NoExecute"
 ```
 
+#### CPU Architecture awareness for single-arch clusters 
+
+Users of single-arch non-amd64 clusters who are using scale from zero 
+support should also set the `CAPI_SCALE_ZERO_DEFAULT_ARCH` environment variable
+to the architecture of the nodes they want to default the node group templates to.
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/README.md:284]: minor nit, missing a word

```suggestion
to set the architecture of the nodes they want to default the node group templates to.
```

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils.go [hunk 6]
```diff
@@ -279,3 +311,37 @@ func getClusterNameLabel() string {
...
+	}
+}
+
+// GetDefaultScaleFromZeroArchitecture returns the SystemArchitecture from the environment variable
+// CAPI_SCALE_ZERO_DEFAULT_ARCH or DefaultArch if the variable is set to an invalid value.
+func GetDefaultScaleFromZeroArchitecture() SystemArchitecture {
+	once.Do(func() {
+		archStr := os.Getenv(scaleUpFromZeroDefaultArchEnvVar)
+		arch := SystemArchitectureFromString(archStr)
+		klog.V(5).Infof("the  value is set to %s (%s)", scaleUpFromZeroDefaultArchEnvVar, archStr, arch.Name())
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils.go:338]: i think we should add little extra detail to this log

```suggestion
		klog.V(5).Infof("the default scale from zero architecture value is set to %s (%s)", scaleUpFromZeroDefaultArchEnvVar, archStr, arch.Name())
```

**Review by elmiko — COMMENTED** [review]
2 minor nits them i'm ready to merge @aleskandro

**Review by elmiko — COMMENTED** [review]
thanks @aleskandro 

/approve
/lgtm
